tensorflow_io/core/python/experimental/avro_dataset_ops.py

```diff
@@ -0,0 +1,627 @@
+                wildcard '*' to indicate that we want to select all 'first_name'
+                entries from the array.
+
+                features = {'sparse_features':
+                            tf.SparseFeature(index_key='index',
+                                             value_key='value',
+                                             dtype=tf.float32, size=10)}
+
+                We assume that sparse features contains an array with records
+                that contain an 'index' field that MUST BE LONG and an 'value'
+                field with floats (single precision).
+
+      batch_size: Items in a batch, must be > 0
+
+      num_parallel_calls: Number of parallel calls
+
+      label_key: The label key, if None no label will be returned
+
+      num_epochs: The number of epochs. If number of epochs is set to None we
+                  cycle infinite times and drop the remainder automatically.
+                  This will make all batch sizes the same size and static.
+
+      input_stream_buffer_size: The size of the input stream buffer in By
+
+      avro_data_buffer_size: The size of the avro data buffer in By
+
+    """
+    dataset = tf.data.Dataset.from_tensor_slices(filenames)
+    if shuffle:
+        n_filenames = tf.shape(filenames, out_type=tf.int64)[0]
+        dataset = dataset.shuffle(n_filenames, shuffle_seed)
+
+    if label_keys is None:
+        label_keys = []
+
+    # Handle the case where the user only provided a single label key
+    if not isinstance(label_keys, list):
+        label_keys = [label_keys]
+
+    for label_key in label_keys:
+        if label_key not in features:
+            raise ValueError(
+                "`label_key` provided (%r) must be in `features`." % label_key
+            )
+
+    def filename_to_dataset(filename):
+        # Batches
+        return _AvroDataset(
+            filenames=filename,
+            features=features,
+            reader_schema=reader_schema,
+            batch_size=batch_size,
+            drop_remainder=num_epochs is None,
+            num_parallel_calls=num_parallel_calls,
+            input_stream_buffer_size=input_stream_buffer_size,
+            avro_data_buffer_size=avro_data_buffer_size,
+        )
+
+    # Read files sequentially (if num_parallel_reads=1) or in parallel
+    dataset = dataset.interleave(
+        filename_to_dataset,
+        cycle_length=num_parallel_calls,
+        num_parallel_calls=num_parallel_reads,
+    )
+
+    if shuffle:
+        dataset = dataset.shuffle(shuffle_buffer_size, shuffle_seed)
+    if num_epochs != 1:
+        dataset = dataset.repeat(num_epochs)
+
+    if any(isinstance(feature, tf.io.SparseFeature) for _, feature in features.items()):
+        # pylint: disable=protected-access
+        # pylint: disable=g-long-lambda
+        dataset = dataset.map(
+            lambda x: parsing_ops._construct_sparse_tensors_for_sparse_features(
```

Comment:
This `parsing_ops._construct_sparse_tensors_for_sparse_features` need to be replaced by a public API.

If you take a look at the source code in TF, you probably understand why it is dangerous to use it:
https://github.com/tensorflow/tensorflow/blob/5f931045eefcc4d2e8473fcf694a0c52adb7935f/tensorflow/python/ops/parsing_ops.py#L50-L57

There are too many indirections and it is already as a TOTO to switch the location. 

You can copy:
https://github.com/tensorflow/tensorflow/blob/6d0f422525d8c1dd3184d39494abacd32b52a840/tensorflow/python/ops/parsing_config.py#L661

And you can skip `RaggedFeature` part for now.

Reply:
thanks, updated.